dali/kernels/util.h

```diff
@@ -114,7 +114,69 @@ template <typename T>\
 struct has_unique_function_##function_name : \
   decltype(HasUniqueFunction_##function_name<T>(nullptr)) {}; \
 
-}  // namespace kernels
+
+/// @brief Type of a volume with given `ExtentType`
+template <typename ExtentType, bool arithm = std::is_arithmetic<ExtentType>::value>
+struct volume_type {
+  using type = decltype(std::declval<ExtentType>() * std::declval<ExtentType>());
+};
+
+/// @brief volume_type is undefined for non-arithmetic types, by default
+template <typename ExtentType>
+struct volume_type<ExtentType, false> {};
+
+/// @brief 32-bit integers need promotion to 64-bit to store volume safely
+template <>
+struct volume_type<int32_t, true> {
+  using type = int64_t;
+};
+
+/// @brief 32-bit integers need promotion to 64-bit to store volume safely
+template <>
+struct volume_type<uint32_t, true> {
+  using type = uint64_t;
+};
+
+template <typename ExtentType>
+using volume_t = typename volume_type<
+  typename std::remove_const<typename std::remove_reference<ExtentType>::type>::type>::type;
+
+/// @brief Returns the product of all elements in shape
+/// @param shape_begin - start of the shape extent list
+/// @param shape_end - end of the shape extent list
+template <typename Iter>
+inline volume_t<decltype(*std::declval<Iter>())>
+volume(Iter shape_begin, Iter shape_end) {
+  if (shape_begin == shape_end)
+    return 0;  // perhaps we should return 1 as a neutral element of multiplication?
```

Comment:
But then shapeless tensors will return 1 element.

Reply:
I was wondering at some point if we should have a notion of 0-d tensor - which would then be a glorified scalar. It would fit nicely into unifying support ops (they would produce a list of 0d tensors, which might be specialized for more efficient representation). But maybe that's an overkill.